### ISSUE TITLE
Multiple records of procedure 

### ISSUE THREAD

**Issue by cdtv0833** [rana/ora#85]
I use ora like this 

      id := make([]int64, 1)
	for k, _ := range id {
		id[k] = int64(k + 1)
	}
	names := make([]string, 1)
	for k1, _ := range names {
		names[k1] = "a"
	}
--after oracle conect
stmt, err := ses.Prep("begin test(:1,:2);end;")
_, err = stmt.Exe(id, names)
there is a error
PLS-00306: wrong number or types of arguments in call to 'test';

If i use single mod,It's OK.

The multiple mod is  not   supported procedure ?

**Comment by rana** [comment]
Can you include your proc code?

**Comment by cdtv0833** [comment]
this is my proc 

CREATE OR REPLACE PROCEDURE test(p_id in number,p_name in varchar2) 
AS
BEGIN
 insert into test(id,name) values(p_id,p_name);
END;
/
 thank you

**Comment by tgulacsi** [comment]
If you have only this INSERT in your procedure, then you can use that directly in the ses.Exe:

    ses.PrepAndExe("INSERT INTO test(id,name) VALUES (:1, :2)", id, names)

Otherwise, your procedure should accept some array type:

```
CREATE OR REPLACE PACKAGE test AS
  TYPE num_tab_typ IS TABLE OF NUMBER INDEX BY PLS_INTEGER;
  TYPE vc30_tab_typ IS TABLE OF VARCHAR2(30) INDEX BY PLS_INTEGER;
  PROCEDURE t1(p_id IN num_tab_typ, p_name IN vc30_tab_typ);
END;
/
CREATE OR REPLACE PACKAGE BODY test AS
 PROCEDURE t1(p_id IN num_tab_typ, p_name IN vc30_tab_typ) IS
  v_idx PLS_INTEGER;
 BEGIN
  v_idx := p_id.FIRST;
  WHILE v_idx IS NOT NULL LOOP
    DBMS_OUTPUT.PUT_LINE(v_idx||': '||p_id(v_idx)||'='||p_name(v_idx));
    v_idx := p_id.NEXT(v_idx);
  END LOOP;
 END;
END;
/
```

**Comment by cdtv0833** [comment]
I just give an example.the silce len is not 1.
My procedure has been build.  My procedure is complicated  to example .I want to use Multiple Mod to use procedure. In python ,
I can use cursor.executemany to finish it.

**Comment by tgulacsi** [comment]
What does `ses.PrepAndExe("INSERT INTO test(id,name) VALUES (:1, :2)", id, names)` say?

**Comment by cdtv0833** [comment]
ses.PrepAndExe("INSERT INTO test(id,name) VALUES (:1, :2)", id, names) is ok.
It's completed and insert database multiple data.
ses.PrepAndExe("begin  test(:1,:2);end;", id, names) is not ok with error;the error is
PLS-00306: wrong number or types of arguments in call to 'QWTEST'

**Comment by tgulacsi** [comment]
Yup, pl/sql needs a proper table type as I shown

**Comment by cdtv0833** [comment]
@tgulacsi  Your method can solve my problem.Thank you very much.
But I still have a question.
Same multiple mod to exec procedure, with python , I can use cursor.executemany to finish it. with GO It's not completed. That's why?

**Comment by tgulacsi** [comment]
Hmm.

It's because I wanted to be able to call stored procedures with arrays. Ora only allowed calling like executemany. When I introduced calling sp with slices, I had to decide which calling mode to use in Exe.
Not it's done by checking the statement's type: if it's a PL/SQL block (starts with DECLARE/BEGIN), then calls with the slices, otherwise like executemany.

I see several possibilities, with different disadvantages:
a) add ExeM (as executemany) to call as executemany, modify Exe to alwasy call sp with slice,
b) add ExeS to call sp with slice, and modify Exe to always call as executemany (old, consistent behaviour),
c) make Exe's behaviour configurable with StmtCfg.

a) is a rude backward-incompatible change,
b) is backwards-incompatible, but call with slice is a very new feature, I think only I use it,
c) is as backwards-incompatible as b) and adds some complexity.

TL;DR; I'd go with b).

**Comment by cdtv0833** [comment]
@tgulacsi  Thank you very much.

**Comment by tgulacsi** [comment]
It's not ready yet!

I'd like to get some feedback from Rana first, then it has to be coded...

**Comment by rana** [comment]
As you wish, go for it!

**Comment by tgulacsi** [comment]
@cdtv0833 can you try the "issue85" branch?

**Comment by cdtv0833** [comment]
@tgulacsi    It's not completed with the same error

**Comment by tgulacsi** [comment]
Please take another look - added a test (example) and now it passes for me.

**Comment by cdtv0833** [comment]
It's completed with the  not error ! thank you !
